front_end/src/utils/charts.ts

```diff
@@ -522,56 +589,125 @@ export function generateScale({
 export function generateChoiceItemsFromMultipleChoiceForecast(
   question: QuestionWithMultipleChoiceForecasts,
   config?: {
+    withMinMax?: boolean;
     activeCount?: number;
+    preselectedQuestionId?: number;
+    locale?: string;
+    preserveOrder?: boolean;
   }
 ): ChoiceItem[] {
-  const { activeCount } = config ?? {};
+  const { activeCount, preserveOrder } = config ?? {};
 
   const latest = question.aggregations.recency_weighted.latest;
+  if (!latest) {
+    return [];
+  }
 
   const choiceOrdering: number[] = question.options!.map((_, i) => i);
-  choiceOrdering.sort((a, b) => {
-    const aCenter = latest?.forecast_values[a] ?? 0;
-    const bCenter = latest?.forecast_values[b] ?? 0;
-    return bCenter - aCenter;
+  if (!preserveOrder) {
+    choiceOrdering.sort((a, b) => {
+      const aCenter = latest?.forecast_values[a] ?? 0;
+      const bCenter = latest?.forecast_values[b] ?? 0;
+      return bCenter - aCenter;
+    });
+  }
+
+  const labels = question.options!;
+  const aggregationHistory = question.aggregations.recency_weighted.history;
+  const userHistory = question.my_forecasts?.history;
+
+  const aggregationTimestamps: number[] = [];
+  aggregationHistory.forEach((forecast) => {
+    aggregationTimestamps.push(forecast.start_time);
+    if (forecast.end_time) {
+      aggregationTimestamps.push(forecast.end_time);
+    }
+  });
+  const sortedAggregationTimestamps = uniq(aggregationTimestamps).sort(
+    (a, b) => a - b
+  );
+  const userTimestamps: number[] = [];
+  userHistory?.forEach((forecast) => {
+    userTimestamps.push(forecast.start_time);
+    if (forecast.end_time) {
+      userTimestamps.push(forecast.end_time);
+    }
   });
+  const sortedUserTimestamps = uniq(userTimestamps).sort((a, b) => a - b);
```

Comment:
A bunch of code here is duplicated in `generateChoiceItemsFromGroupQuestions`. Can we reduce code duplication by introducing a reusable helper?

Reply:
I gave it a try, but they are different enough that factoring out to a helper function won't help very much.
I have given it more thought and I actually think it might be worth making it so the Question my_forecasts and aggregations.<aggregation_name> histories are already in this format, meaning no data mangling would be required in the front end.

I think due to the intent on more serious refactoring later, as well as my genuine attempt at refactoring now not working out, I'm going to keep it this way for the time being.